server/src/main/java/com/linecorp/centraldogma/server/TlsConfig.java

```diff
@@ -15,55 +15,117 @@
  */
 package com.linecorp.centraldogma.server;
 
-import static java.util.Objects.requireNonNull;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.linecorp.centraldogma.server.CentralDogmaConfig.convertValue;
 
+import java.io.ByteArrayInputStream;
 import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
 
 import javax.annotation.Nullable;
 
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.common.base.MoreObjects;
+import com.google.errorprone.annotations.MustBeClosed;
 
 /**
  * TLS configuration.
  */
 public final class TlsConfig {
 
+    @Nullable
     private final File keyCertChainFile;
+    @Nullable
     private final File keyFile;
     @Nullable
+    private final String keyCertChain;
+    @Nullable
+    private final String key;
+    @Nullable
     private final String keyPassword;
 
     /**
      * Creates an instance with the specified {@code keyCertChainFilePath}, {@code keyFilePath} and
      * {@code keyPassword}.
      */
     @JsonCreator
-    public TlsConfig(@JsonProperty(value = "keyCertChainFile", required = true) File keyCertChainFile,
-                     @JsonProperty(value = "keyFile", required = true) File keyFile,
+    public TlsConfig(@JsonProperty(value = "keyCertChainFile") @Nullable File keyCertChainFile,
+                     @JsonProperty(value = "keyFile") @Nullable File keyFile,
+                     @JsonProperty(value = "keyCertChain") @Nullable String keyCertChain,
+                     @JsonProperty(value = "key") @Nullable String key,
                      @JsonProperty("keyPassword") @Nullable String keyPassword) {
-        this.keyCertChainFile = requireNonNull(keyCertChainFile, "keyCertChainFile");
-        this.keyFile = requireNonNull(keyFile, "keyFile");
+        checkArgument(keyCertChainFile != null || keyCertChain != null,
+                      "keyCertChainFile and keyCertChain cannot be null at the same time.");
+        checkArgument(keyFile != null || key != null,
+                      "keyFile and key cannot be null at the same time.");
+        this.keyCertChainFile = keyCertChainFile;
+        this.keyFile = keyFile;
+        this.keyCertChain = keyCertChain;
+        this.key = key;
         this.keyPassword = keyPassword;
     }
 
     /**
      * Returns a certificates file which is created with the given {@code keyCertChainFilePath}.
+     *
+     * @deprecated Use {@link #keyCertChainInputStream()}.
      */
+    @Nullable
     @JsonProperty
+    @Deprecated
     public File keyCertChainFile() {
         return keyCertChainFile;
     }
 
     /**
      * Returns a private key file which is created with the given {@code keyFilePath}.
+     *
+     * @deprecated Use {@link #keyInputStream()}.
      */
+    @Nullable
     @JsonProperty
+    @Deprecated
```

Comment:
Note: I was originally confused since I thought this didn't need to be deprecated (e.g. it may be useful to read an encrypted file and decrypt on-the-fly)

Reply:
That's what the converter is for. 😉
```
"tls.key" : "encryption:myKeyVariable"

class NClavisConvigValueConverter implements ConfigValueConverter {
  public string convert(String prefix, String value) {
    assert prefix.equals("encryption");
    if ("myKeyVariable".equals(value)) {
      // Read a specific encrypted file and decrypt it using thrid-party service.
    }
    ...
  }
}
```